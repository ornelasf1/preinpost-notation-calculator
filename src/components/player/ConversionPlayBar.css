
@media screen and (max-width: 1200px) {
  #progressBar #playBtn {
    width: 40px;
    height: 40px;
  }

  .toolbarBtn {
    width: 30%;
    height: 50%;
    margin: 10px 20px;
  }

  .toolbarBtn > * {
    width: 20%;
    height: 20px;
  }
}

#playBtn {
    width: 50px;
    height: 50px;
}

#playerToolbar {
  height: 60px;
  background-color: var(--lightcolor3);
  display: flex;
  justify-content: center;
}

.toolbarBtn {
  width: 25%;
  margin: 10px 20px;
}

.toolbarBtn > * {
  width: 25%;
  height: 25px;
}

.disableBtn {
  background-color: var(--lightcolor2);;
}