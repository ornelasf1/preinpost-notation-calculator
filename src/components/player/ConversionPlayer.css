
@media screen and (max-width: 1200px) {
    #tokens {
        min-width: 250px;
        height: 40px;
    }

    .inputtoken {
        font-size: 30px;
    }

    #precTable {
        transform: scale(0.7);
    }
}

.player{
    background-color: white;
    height: 100%;
    flex-direction: column;
    text-align: center;

    /* Position relative in order to have children to be absolute positioned */
    position: relative;
}

.contentBox {
    white-space: nowrap;
    overflow-x: hidden;
    overflow-y: hidden;
    padding-bottom: 10px;
    overflow: visible;
}


#tokens {
    /* Position is relative to calculate position for token cursor */
    position: relative;  
    margin: auto;
    margin-top: 20px;
    width: fit-content;
    padding-left: 50px;
    padding-right: 50px;
    max-width: 800px;
    min-width: 300px;
    height: 50px;
    background-color: lightgray;
    border-radius: 10px;
}


.inputtoken {
    display: inline-block;
    font-size: 35px;
    font-weight: 600;
    width: auto;
    margin-left: 10px;
}

#tokenCursor {
    background: url(../../imgs/chevron-arrow-up.svg);
    background-repeat: no-repeat;
    background-size: 24px 24px;
    position: relative;
    height: inherit;
    top: 0;
    right: 0;
    opacity: 0;
    transition-property: left, opacity;
    transition: 1s;
    animation: hoverCursor 1s ease-in-out;
    animation-iteration-count: infinite;
}

.animatedToken {
    transition-property: left, top;
    transition: 1s;
}

@keyframes hoverCursor {
    0%, 100% {
        transform: translateY(-5px);
    }
    50% {
        transform: translateY(0px);
    }
}

#structures {
    margin-top: 80px;
    height: auto;
    display: inline-flex;
}

#stack {
    width: 50px;
    min-height: 140px;
    height: auto;
    margin-top: auto;
    border: 5px solid black;
    border-top-color: white;
    margin-right: 20px;
    display: flex;
    flex-direction: column-reverse;
    transition-property: height;
    transition: 1s ease-in-out;
}

.stacktoken {
    font-size: 35px;
    font-weight: 600;
    position: relative;
}

#output {
    display: inline-block;
    margin-top: 150px;
    width: auto;
    min-width: 300px;
    max-width: 800px;
    height: 50px;
    border-bottom: 5px solid black;
    border-radius: 10px;
    margin-left: 20px;
}

.outputtoken {
    display: inline-block;
    font-size: 35px;
    font-weight: 600;
    width: auto;
    padding-left: 10px;
}

.activeStructure {
    animation: glow .5s alternate infinite ease-in-out;
}
@keyframes glow {
    from {
        box-shadow: 0 0 3mm 3px var(--activity-color);
    }
    to {
        box-shadow: 0 0 5mm 5px var(--activity-color)
    }
}



#helpWindow {
    background-color: rgb(211, 211, 211);
    border-radius: 10px;
    margin: 20px;
    padding: 20px;
    width: auto;
    min-height: 60px;
    opacity: 0;
    transition: opacity .25s ease-in-out;
}

#helpWindow > div {
    font-size: 30px;
}

.helpItem {
    display: inline;
    font-size: 40px;
    margin: 0 5px;
    font-weight: bold;
    position: relative;
    cursor: pointer;
}

.helpText > span {
    display: block;
    font-size: 20px;
}

.helpItem .helpText {
    opacity: 0;
    text-align: center;
    position: absolute;
    z-index: 1;
    background-color: rgb(228, 228, 228);
    border-radius: 10px;
    padding: 5px 10px;
    left: -20px;
    top: -160%;
    transition: opacity .1s ease-in-out;
}

.helpItem:hover .helpText{
    opacity: .8;
}



#precTable {
    background-color: white;
    position: absolute;
    left: 80%;
    top: 20px;
}

th {
    padding: 10px;
    background-color: var(--primary-color);
    color: white;
}

td {
    padding-bottom: 8px;
    color: white;
    font-size: 30px;
}
td:nth-child(odd) {
    font-weight: bold;
}

tr:nth-child(odd) {
    background-color: var(--instruction-color-1);
}
tr:nth-child(even) {
    background-color: var(--instruction-color-2);
}