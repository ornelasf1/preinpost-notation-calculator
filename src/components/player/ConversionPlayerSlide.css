@media screen and (max-width: 1200px) {
    #progressBar #playBtn {
      width: 40px;
      height: 40px;
    }

    .slideContainer .slider::-webkit-slider-thumb {
        width: 8px;
        height: 40px;
    }
    
    .slideContainer .slider::-moz-range-thumb {
        width: 8px;
        height: 40px;
    }
}

#progressBar{
    display: flex;
}

#playBtn {
    margin-right: 5px;
    margin-bottom: 5px;
    border-radius: 0px;
    border-top-left-radius: 15px;
}

.slideContainer {
    width: 100%;
    margin-bottom: 5px;
}

.slider {
    -webkit-appearance: none;
    width: 100%;
    background-color: var(--alt-color);
    outline: none;
    border-radius: 0px;
    border-top-right-radius: 15px;

}

.slider::-webkit-slider-thumb {
    -webkit-appearance: none;
    appearance: none;
    width: 10px;
    height: 50px;
    background: var(--lightcolor1);
    cursor: pointer;
    border-radius: 3px;
}

.slider::-moz-range-thumb {
    -webkit-appearance: none;
    appearance: none;
    width: 10px;
    height: 50px;
    background: var(--lightcolor1);
    cursor: pointer;
    border-radius: 3px;
}