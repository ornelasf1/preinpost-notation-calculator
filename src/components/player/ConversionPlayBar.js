
import React from 'react';
import {connect} from 'react-redux';
import Slider from 'react-input-slider';
import './ConversionPlayBar.css';
import {ReactComponent as PlayIcon} from '../../play.svg';
import {ReactComponent as PauseIcon} from '../../pause.svg';
import {ReactComponent as ForwardIcon} from '../../../src/imgs/forward_arrow.svg';
import {ReactComponent as RewindIcon} from '../../../src/imgs/rewind_arrow.svg';

class ConversionPlayBar extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            isPlaying: false,
            instructionIndex: -1,
            disableRewind: true,
            disableForward: false,
        }
        this.playSeq = undefined;
    }

    componentDidUpdate = (prevProps) => {
        const { selectedNotation, toNotation } = this.props;
        if (prevProps.conversion.expressions[selectedNotation]
            !== this.props.conversion.expressions[selectedNotation]) {
                this.resetPlayer();
            }

        const instructionSequence = this.getInstructionSequence(selectedNotation, toNotation);
        const newInstructionIndex = this.state.instructionIndex < instructionSequence.length?
            this.state.instructionIndex : 0;
        this.props.updateSelectedInstruction(newInstructionIndex);
    }

    resetPlayer = () => {
        this.playSeq = undefined;
        this.setState({
            isPlaying: false,
            instructionIndex: -1,
        });
    }

    getInstructionSequence = (fromNotation, toNotation) => {
        const instructionSetName = fromNotation + 'Instr';
        const notationSequences = this.props.algorithm[instructionSetName];

        if (toNotation === 'prefix') {
            return notationSequences.toPreInstr;
        }else if (toNotation === 'infix') {
            return notationSequences.toInInstr;
        }else if (toNotation === 'postfix') {
            return notationSequences.toPostInstr;
        }
    }

    handlePlayBtn = () => {
        const { selectedNotation, toNotation } = this.props;
        const instructionSequence = this.getInstructionSequence(selectedNotation, toNotation);

        this.setState({
            isPlaying: !this.state.isPlaying,
            disableRewind: this.state.instructionIndex === -1,
        });

        if (this.playSeq === undefined) {
            console.log('Initiated');
            this.playSeq = new Timer(() => {
                console.log('RUNNING');
                this.setState({
                    instructionIndex: this.state.instructionIndex + 1,
                    disableRewind: this.state.instructionIndex === -1,
                }, () => {
                    if (this.playSeq !== undefined) {
                        this.playSeq.setIndex(this.state.instructionIndex);
                        console.log('Compare outside: ', this.playSeq.getIndex(), instructionSequence.length);
                        if (this.playSeq.getIndex() === instructionSequence.length) {
                            console.log('reset');
                            this.setState({
                                isPlaying: false, 
                                instructionIndex: -1,
                                disableRewind: true,
                                disableForward: false,
                            });
                            this.playSeq = undefined;
                        }

                    }
                });
            }, instructionSequence.length, this.state.instructionIndex, 500);

        } else {
            if (this.state.isPlaying) {
                console.log('pausing')
                this.playSeq.pause();
            } else {
                console.log('playing')
                this.playSeq.resume();
            }
        }
    }

    handleForwardBtn = () => {
        const { selectedNotation, toNotation} = this.props;
        const instructionSequence = this.getInstructionSequence(selectedNotation, toNotation);
        const instructionSequenceLimit = instructionSequence.length;
        const newInstructionIndex = this.state.instructionIndex + 1 < instructionSequenceLimit?
            this.state.instructionIndex + 1 : this.state.instructionIndex;
        this.setState({
            isPlaying: false,
            disableForward: newInstructionIndex + 1 === instructionSequenceLimit,
            disableRewind: false,
            instructionIndex: newInstructionIndex,
        });
    }

    handleRewindBtn = () => {
        const newInstructionIndex = this.state.instructionIndex - 1 <= -1? -1: this.state.instructionIndex - 1;
        this.setState({
            isPlaying: false,
            disableRewind: newInstructionIndex === -1,
            disableForward: false,
            instructionIndex: newInstructionIndex,
        });
    }

    render = () => {
        const { isPlaying, disableForward, disableRewind } = this.state;
        const { selectedNotation, toNotation,  conversion: { valid }} = this.props;
        const instructionSequence = this.getInstructionSequence(selectedNotation, toNotation);
        if (this.playSeq && !isPlaying) {
            this.playSeq.pause();
        }

        return (
            <div>
                <div id='progressBar'>
                    <button disabled={!valid} id='playBtn' onClick={this.handlePlayBtn}>
                        {isPlaying? <PauseIcon /> : <PlayIcon />}
                    </button>
                    <Slider 
                        id='slider'
                        disabled={!valid}
                        axis="x"
                        xstep={1}
                        xmin={-1}
                        xmax={instructionSequence.length - 1}
                        x={this.state.instructionIndex}
                        onChange={({x}) => {
                            this.setState({
                                isPlaying: false,
                                disableRewind: x === -1,
                                disableForward: x === instructionSequence.length - 1,
                                instructionIndex: x})
                        }}
                        styles={{
                            track: {
                                backgroundColor: '#143c41'
                              },
                              active: {
                                backgroundColor: '#1a484e'
                              },
                              thumb: {
                                width: 10,
                                height: 10,
                                top: '50%',
                              }
                        }}
                        />
                </div>
                <div id="playerToolbar">
                      <button 
                        id='rewindBtn' 
                        className={'toolbarBtn' + (disableRewind? ' disableBtn' : '')}
                        onClick={this.handleRewindBtn} 
                        disabled={disableRewind}>
                          <RewindIcon />
                      </button>
                      <button 
                        id='forwardBtn' 
                        className={'toolbarBtn' + (disableForward? ' disableBtn' : '')}
                        onClick={this.handleForwardBtn} 
                        disabled={disableForward}>
                          <ForwardIcon />
                          </button>  
                </div>
            </div>
        );
    }
}

function Timer(callback, amountOfIntructions, initInstructionIndex, interval) {
    var timerId;
    var instructionIndex = initInstructionIndex;

    this.setIndex = newInstructionIndex => {
        console.log('index updated ', instructionIndex, newInstructionIndex);
        instructionIndex = newInstructionIndex;
    };

    this.getIndex = () => instructionIndex;

    this.pause = () => {
        clearTimeout(timerId);
    };

    this.resume = () => {
        console.log('INTERVAL: ', instructionIndex);
        console.log('Compare: ', instructionIndex, amountOfIntructions)
        if (instructionIndex < amountOfIntructions) {
            callback();
            timerId = setTimeout(() => {
                this.resume();
            }, interval);
        } else {
            console.log('END');
            clearTimeout(timerId);
        }
    };
    this.resume();
}

const mapStateToProps = state => ({
    algorithm: state.algorithmInstructions,
    conversion: state.conversionNotat,
});

export default connect(mapStateToProps)(ConversionPlayBar);