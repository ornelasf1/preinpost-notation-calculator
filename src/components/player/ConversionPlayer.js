
import React from 'react';
import _ from 'underscore';
import './ConversionPlayer.css';

export class ConversionPlayer extends React.Component {

    constructor(props) {
        super(props);

        this.lastestTopStack = {};
    }

    getSnapshotBeforeUpdate = prevProps => {
        const topStackIndex = prevProps.selectedInstr.stackTokens.length - 1;
        const topStackToken = document.getElementById('stacktoken-'+topStackIndex);
        this.lastestTopStack = {
            left: topStackToken ? topStackToken.offsetLeft : 0,
            top: topStackToken ? topStackToken.offsetTop : 0,
            node: topStackToken,
        };

        const topOutputIndex = prevProps.selectedInstr.outputTokens.length - 1;
        this.lastTopOutputToken = document.getElementById('outputtoken-'+topOutputIndex);
    }

    componentDidUpdate = prevProps => {
        if (prevProps.selectedInstr.selectedTokenIndex !==
            this.props.selectedInstr.selectedTokenIndex 
            || this.props.selectedInstr.selectedTokenIndex === -1) {
                this.updateTokenStyle(this.props.selectedInstr.selectedTokenIndex);
            }
        this.animateTokens(prevProps.selectedInstr, this.props.selectedInstr);
    }

    getCoords = elem => {
        let box = elem.getBoundingClientRect();
        return {
            top: box.top + window.pageYOffset,
            left: box.left + window.pageXOffset,
            height: box.height,
        };
    }

    updateTokenStyle = index => {
        const tokenCursor = document.getElementById('tokenCursor');
        if (tokenCursor === undefined) {
            return;
        }
        if (index !== -1 && index < this.props.expression.length) {
            const tokenElem = document.getElementById('inputtoken-'+index);
            const tokenParentElemLeftPadding = parseFloat(window.getComputedStyle(tokenElem.parentElement).paddingLeft.replace('px',''));
            const widthOfCursor = parseFloat(getComputedStyle(tokenCursor).backgroundSize.split(' ')[0].replace('px',''));
            tokenCursor.style.left = (tokenElem.offsetLeft + (tokenElem.offsetWidth / 2) - tokenParentElemLeftPadding - (widthOfCursor/2)) + 'px';
            tokenCursor.style.opacity = 1;
        }else {
            tokenCursor.style.left = '0px';
            tokenCursor.style.opacity = 0;
        }
    }

    animateTokens = (prevSelectedInstr, currSelectedInstr) => {
        const playerScreen = document.getElementById('tokens').parentNode;

        if (prevSelectedInstr.stackTokens.length < currSelectedInstr.stackTokens.length) {
            const originLocation = document.getElementById('inputtoken-'+currSelectedInstr.selectedTokenIndex);
            const tokenToMove = originLocation.cloneNode(true);

            tokenToMove.style.position = 'absolute';
            tokenToMove.className += ' animatedToken';

            tokenToMove.style.top  = originLocation.parentNode.offsetTop + 'px';
            tokenToMove.style.left = (originLocation.offsetLeft + 
                                    document.getElementById('tokens').offsetLeft - 
                                    parseFloat(getComputedStyle(originLocation).marginLeft.replace('px',''))) + 'px';
            playerScreen.append(tokenToMove);

            const topIndex = currSelectedInstr.stackTokens.length - 1;
            const topStackToken = document.getElementById('stacktoken-'+topIndex);
            const tokenAnimationDuration = parseFloat(getComputedStyle(tokenToMove).transitionDuration.replace('s',''));
            
            topStackToken.style.opacity = 0;
            tokenToMove.style.top = topStackToken.offsetTop + 'px';
            tokenToMove.style.left = (topStackToken.offsetLeft + 3) + 'px';
            
            setTimeout(() => {
                topStackToken.style.opacity = 1;
                tokenToMove.remove();
                console.log('delete temp token and reappear stack token');
            }, tokenAnimationDuration * 1000);
        } else if (prevSelectedInstr.stackTokens.length - 1 === currSelectedInstr.stackTokens.length &&
            prevSelectedInstr.outputTokens.length === currSelectedInstr.outputTokens.length - 1) {
                const originLocation = this.lastestTopStack.node;
                const tokenToMove = originLocation.cloneNode(true);

                tokenToMove.style.position = 'absolute';
                tokenToMove.className += ' animatedToken';

                tokenToMove.style.top  = this.lastestTopStack.top + 'px';
                tokenToMove.style.left = (this.lastestTopStack.left + 3) + 'px';
                playerScreen.append(tokenToMove);

                const topOutputIndex = currSelectedInstr.outputTokens.length - 1;
                const topOutputToken = document.getElementById('outputtoken-'+topOutputIndex);
                const tokenAnimationDuration = parseFloat(getComputedStyle(tokenToMove).transitionDuration.replace('s',''));

                topOutputToken.style.opacity = 0;
                tokenToMove.style.top = topOutputToken.offsetTop + 'px';
                tokenToMove.style.left = (topOutputToken.offsetLeft + parseFloat(getComputedStyle(topOutputToken).marginLeft.replace('px',''))) + 'px';
                
                setTimeout(() => {
                    topOutputToken.style.opacity = 1;
                    tokenToMove.remove();
                    console.log('delete temp token and reappear stack token');
                }, tokenAnimationDuration * 1000);
        } else if (prevSelectedInstr.outputTokens.length < currSelectedInstr.outputTokens.length) {
            const originLocation = document.getElementById('inputtoken-'+currSelectedInstr.selectedTokenIndex);
            const tokenToMove = originLocation.cloneNode(true);

            tokenToMove.style.position = 'absolute';
            tokenToMove.className += ' animatedToken';

            tokenToMove.style.top  = originLocation.parentNode.offsetTop + 'px';
            tokenToMove.style.left = (originLocation.offsetLeft + 
                                    document.getElementById('tokens').offsetLeft - 
                                    parseFloat(getComputedStyle(originLocation).marginLeft.replace('px',''))) + 'px';
            playerScreen.append(tokenToMove);

            const topIndex = currSelectedInstr.outputTokens.length - 1;
            const topStackToken = document.getElementById('outputtoken-'+topIndex);
            const tokenAnimationDuration = parseFloat(getComputedStyle(tokenToMove).transitionDuration.replace('s',''));
            
            topStackToken.style.opacity = 0;
            tokenToMove.style.top = topStackToken.offsetTop + 'px';
            tokenToMove.style.left = topStackToken.offsetLeft + 'px';
            
            setTimeout(() => {
                topStackToken.style.opacity = 1;
                tokenToMove.remove();
                console.log('delete temp token and reappear stack token');
            }, tokenAnimationDuration * 1000);
        }
    }

    render = () => {
        const { selectedInstr, expression } = this.props;

        //Expects all expressions in state to be demilimited by spaces
        const expressionTokens = expression.split(' ');
        const divInputTokens = expressionTokens.map((token, idx) => <div key={idx} id={"inputtoken-"+idx} className='inputtoken'>{token}</div>);
        const divStackTokens = selectedInstr.stackTokens.map((token, idx) => <div key={idx} id={"stacktoken-"+idx} className='stacktoken'>{token}</div>);
        const divOutputTokens = selectedInstr.outputTokens.map((token, idx) => <div key={idx} id={"outputtoken-"+idx} className='outputtoken'>{token}</div>);

        return (
            <div className='player'>
                <div id='tokens' className='contentBox'>
                    {divInputTokens}
                    <div id='tokenCursor'></div>
                </div>
                <div id='structures'>
                    <div id='stack'>
                        {divStackTokens}
                    </div>
                    <div id='output'>
                        {divOutputTokens}
                    </div>
                </div>
            </div>
        );
    };
}
