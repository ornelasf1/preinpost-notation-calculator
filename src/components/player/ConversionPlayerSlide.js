
import React from 'react';
import {ReactComponent as PlayIcon} from '../../imgs/play.svg';
import {ReactComponent as PauseIcon} from '../../imgs/pause.svg';

import './ConversionPlayerSlide.css';

export const ConversionPlayerSlide = props => {
    return (
        <div id='progressBar'>
            <button disabled={props.disabled} id='playBtn' onClick={props.handlePlayBtn}>
                {props.isPlaying? <PauseIcon /> : <PlayIcon />}
            </button>
            <div className="slideContainer">
                <input 
                    className="slider"
                    type="range"
                    disabled={props.disabled}
                    min={props.min} 
                    max={props.max}
                    value={props.value}
                    onChange={event => {
                        props.onChange(event.target.value);
                        console.log(`Change to: ${event.target.value}`);
                        }}>
                </input>
            </div>
        </div>
    );
}