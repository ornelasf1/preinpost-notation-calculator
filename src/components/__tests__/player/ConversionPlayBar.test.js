
import React from 'react';
import { shallow, mount } from 'enzyme';
import {ConversionPlayBar, Timer} from '../../player/ConversionPlayBar';

describe('ConversionPlayBar', () => {
    let wrapper;
    let instance;

    const mockConversionFn = {
        expressions: {
            infix: '',
            prefix: '',
            postfix: '',
        }
    };
    const mockAlgorithmFn = jest.fn();
    const mockUpdateSelectedInstruction = jest.fn();
    const mockSelectedNotation = 'infix';
    const mockToNotation = 'postfix';
    const mockInstructionSequenceLimit = 10;
    const mockInstructionIndex = 4;

    beforeEach(() => {
        wrapper = shallow(<ConversionPlayBar 
            conversion={mockConversionFn} 
            algorithm={mockAlgorithmFn}
            updateSelectedInstruction={mockUpdateSelectedInstruction}
            selectedNotation={mockSelectedNotation}
            toNotation={mockToNotation}
            instructionSequenceLimit={mockInstructionSequenceLimit}/>);

        instance = wrapper.instance();
    });

    it('should render without throwing errors', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should call updateSelectedInstruction on componentDidUpdate', () => {
        wrapper.setProps({ valid: true });
        expect(mockUpdateSelectedInstruction).toBeCalled();
    });

    it('should update instruction index if it\'s below the instruction sequence limit', () => {
        wrapper.setState({instructionIndex: 4});

        expect(mockUpdateSelectedInstruction).toHaveBeenCalledWith(
            instance.state.instructionIndex, 
            mockInstructionSequenceLimit);
        });
        
    it('should update instruction index to 0 if it\'s equal or greater than the sequence limit', () => {
        wrapper.setState({instructionIndex: 10});

        expect(mockUpdateSelectedInstruction).toHaveBeenCalledWith(
            0, 
            mockInstructionSequenceLimit);
    });

    it('should reset player if expression was changed', () => {
        instance.resetPlayer = jest.fn();
        
        wrapper.setProps({conversion: {expressions: {
            [mockSelectedNotation]: '4+3'
        }}});
        
        expect(instance.resetPlayer).toHaveBeenCalled();
    });



    
    it('should reset player if the expression is invalid and the player is active', () => {
        instance.resetPlayer = jest.fn();
        instance.playSeq = new Timer(
            jest.fn(), mockInstructionSequenceLimit, mockInstructionIndex
            );
            wrapper.setProps({ valid: false });
            
            expect(instance.resetPlayer).toHaveBeenCalled();
        });
        
    it("should set playSeq to undefined if it's not already when resetPlayer is called", () => {
        instance.playSeq = new Timer(
            jest.fn(), mockInstructionSequenceLimit, mockInstructionIndex
        );
        instance.setState = jest.fn();
        instance.resetPlayer();

        const resetSeq = instance.setState.mock.calls[0][1];
        resetSeq();

        expect(instance.playSeq).toBe(undefined);
    });

    it("should set playSeqEnded to true if instruction sequence has reached its limit when resetPlayer is called", () => {
        wrapper.setState({instructionIndex: mockInstructionSequenceLimit - 1});
        instance.setState = jest.fn();
        instance.resetPlayer();

        const resetSeq = instance.setState.mock.calls[0][1];
        resetSeq();

        expect(instance.playSeqEnded).toBe(true);
    });

    it("should set playSeqEnded to false if instruction sequence has not reached its limit when resetPlayer is called", () => {
        wrapper.setState({instructionIndex: mockInstructionIndex});
        instance.setState = jest.fn();
        instance.resetPlayer();

        const resetSeq = instance.setState.mock.calls[0][1];
        resetSeq();

        expect(instance.playSeqEnded).toBe(false);
    });

    it("should reset all values to false and -1 when reset player is called with playEnded false", () => {
        instance.resetPlayer();
        
        expect(instance.state.isPlaying).toBe(false);
        expect(instance.state.instructionIndex).toBe(-1);
        expect(instance.state.disableRewind).toBe(false);
        expect(instance.state.disableForward).toBe(false);
    });

    it("should set instruction index to last index in sequence and disable forward on first resetPlayer(true) call, then reset instruction index to -1 and disable rewind on second resetPlayer(true) call", () => {
        instance.resetPlayer(true);
        
        expect(instance.state.isPlaying).toBe(false);
        expect(instance.state.instructionIndex).toBe(mockInstructionSequenceLimit - 1);
        expect(instance.state.disableRewind).toBe(false);
        expect(instance.state.disableForward).toBe(true);

        instance.resetPlayer(true);
        
        expect(instance.state.isPlaying).toBe(false);
        expect(instance.state.instructionIndex).toBe(-1);
        expect(instance.state.disableRewind).toBe(true);
        expect(instance.state.disableForward).toBe(false);
    });



    it("should toggle isPlaying when play button is pressed", () => {
        const isPlaying = instance.state.isPlaying;
        instance.handlePlayBtn();
        
        expect(instance.state.isPlaying).toBe(!isPlaying);
    });
    
    it("should disable rewind when instruction index is -1", () => {
        instance.setState({instructionIndex: -1});
        instance.handlePlayBtn();

        expect(instance.state.disableRewind).toBe(true);
    });

    it("should toggle isPlaying when play button is pressed", () => {
        const isPlaying = instance.state.isPlaying;
        instance.handlePlayBtn();

        expect(instance.state.isPlaying).toBe(!isPlaying);
    });

    it("should toggle resume/pause function of player when playSeq is not undefined", () => {
        instance.playSeq = new Timer(
            jest.fn(), mockInstructionSequenceLimit, mockInstructionIndex
        );
        instance.playSeq.pause = jest.fn();
        instance.playSeq.resume = jest.fn();
        instance.resetPlayer = jest.fn();
        instance.setState({isPlaying: true});

        instance.handlePlayBtn();
        expect(instance.playSeq.resume).toBeCalled();

        instance.handlePlayBtn();
        expect(instance.playSeq.pause).toBeCalled();
    });

    it("should ");
});