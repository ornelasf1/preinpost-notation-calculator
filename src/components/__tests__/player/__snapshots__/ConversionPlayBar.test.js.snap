// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionPlayBar should display play icon when isPlaying is set to false 1`] = `ShallowWrapper {}`;

exports[`ConversionPlayBar should render without throwing errors 1`] = `ShallowWrapper {}`;
